akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -1156,15 +1157,20 @@ private[stream] object Collect {
 
       @tailrec private def pushOne(): Unit =
         if (buffer.isEmpty) {
-          if (isClosed(in)) completeStage()
-          else if (!hasBeenPulled(in)) pull(in)
+          if (isClosed(in)) {
+            if (todo == 0) completeStage()
+          } else if (!hasBeenPulled(in)) pull(in)
         } else if (buffer.peek().elem == NotYetThere) {
           if (todo < parallelism && !hasBeenPulled(in)) tryPull(in)
         } else buffer.dequeue().elem match {
           case Success(elem) ⇒
             push(out, elem)
             if (todo < parallelism && !hasBeenPulled(in)) tryPull(in)
-          case Failure(ex) ⇒ pushOne()
+          case Failure(e) ⇒
+            decider(e) match {
```

Comment:
I tried doing that and it becomes very messy... :(
For one thing, this method is tailrec, due to the pushOne() we would have to drop that then.
Making it common becomes pretty messy really quick, spent quite some time on it now :\

WDYT, should I propose the change or we leave it as is? (becomes pretty complex :/)